src/Geometry.cpp

```diff
@@ -16,510 +14,744 @@
+
+    return dim;
+}
+
+double** Geometry::hexagonoffset(double side) {
+    // pic on top
+    double *dim = dimensions(side);
+    return new double*[2] {
+      new double[6] {dim[1], dim[1], 0., -dim[1],-dim[1], 0},
+      new double[6] {-dim[2], dim[2], side, dim[2], -dim[2],-side}
+    };
+
+    // Hexagons flat side on top
+    // const array<double, nvertices> hexagonoffsetx = {{aover2_,a_,aover2_,-aover2_,-a,-aover2_}};
+    // const array<double, nvertices> hexagonoffsety = {{-asqrt3over2_,0,asqrt3over2_,asqrt3over2_,0.,-asqrt3over2_}};
+}
+
+double** Geometry::triangleoffset(double side){
+
+    // up, up, down, down
+    double *dim = dimensions(side);
+    return new double*[4] {
+      new double[3] {dim[2], 0., -dim[2]},
+      new double[3] {-dim[1]/3., dim[0]/3., -dim[1]/3.},
+      new double[3] {dim[2], -dim[2], 0.},
+      new double[3] {dim[1]/3., dim[1]/3., -dim[0]/3.},
+    };
+}
+
+double* Geometry::derivative(double side, Parameters::Geometry::Type itype){
+    double *derivative = new double[3];
+
+    double *dim = dimensions(side);
+    // partial derivatives of x,y vs i,j
+    switch(itype) {
+        case Parameters::Geometry::Type::Hexagons: {
+            // peak on top
+            derivative[0] = dim[0];
+            derivative[1] = dim[1];
+            derivative[2] = dim[3];
+
+            // // flat side on top 
+            // dydj = asqrt3_;
+            // dxdj = asqrt3over2_;
+            // dxdi = a3over2_;
+            break;
+        }
+        case Parameters::Geometry::Type::Triangles: {
+            derivative[0] = dim[2];
+            derivative[1] = dim[2];
+            derivative[2] = dim[1];
+            break;
+        }
+        default:
+            break;
+    };
+    return derivative;
+}
+
+double** Geometry::dxdyFirstZone(double* xs, double* ys) {
+    double* dx = new double[5];
+    double* dy = new double[5];
+
+    for(unsigned i=0; i<5; i++)
+        dx[i] = xs[i+1]-xs[0];
+    for(unsigned i=0; i<5; i++)
+        dy[i] = ys[i+1]-ys[0];
+
+    return new double*[2] {dx, dy};
+}
+
+double** Geometry::dxdySecondZone(double* xs, double* ys) {
+    double* dx = new double[5];
+    double* dy = new double[5];
+
+    dx[0] = xs[2]-xs[3];
+    dx[1] = xs[6]-xs[3];
+    dx[2] = xs[7]-xs[3];
+    dx[3] = xs[5]-xs[3];
+    dx[4] = xs[8]-xs[3];
+
+    dy[0] = ys[2]-ys[3];
+    dy[1] = ys[6]-ys[3];
+    dy[2] = ys[7]-ys[3];
+    dy[3] = ys[5]-ys[3];
+    dy[4] = ys[8]-ys[3];
+
+    return new double*[2] {dx, dy};
+}
+
+int* Geometry::ijWindows(int zone, double* xs, double* ys, double side, Parameters::Geometry::Type itype) {
+    // compute i,j window needed to cover the x,y window
+
+    int *windows = new int[4];    //imin, imax, jmin, jmax
+
+    double *der = derivative(side, itype);
+    double** dxdy = new double*[5];
+    array<double,6> js_first;
+    array<double,6> is_first;
+    array<double,6> js_second;
+    array<double,6> is_second;
+
+    if (zone == 1) {
```

Comment:
What is zone 1? Use a const variable with a meaningful name instead.

Reply:
Fixed